# BioDiversity Project

# Data Cleaning Steps

Step 1: Eliminate Repeats in Observations (df_obs)
group by scientific_name, and park name, take the median value


Step 2: Clean up common names
Use regex to 
Or take the common names with the longer category

# 1) SetUp

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
df_obs = pd.read_csv("/Users/drewpeterson/Documents/Programming/Python/CodeCademy/biodiversity_starter/observations.csv")

In [5]:
df_species = pd.read_csv("/Users/drewpeterson/Documents/Programming/Python/CodeCademy/biodiversity_starter/species_info.csv")

# 2) Exploration

## Species Data

In [6]:
# Glance at Data
df_species.head(3)

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN

In [7]:
df_species.conservation_status.isna().sum()

5633

In [9]:
df_species.scientific_name.nunique()

5541

In [10]:
df_species.conservation_status.value_counts(dropna=False)

conservation_status
NaN                   5633
Species of Concern     161
Endangered              16
Threatened              10
In Recovery              4
Name: count, dtype: int64

In [12]:
# Summary of Data
    # 23,000 rows
    # 880 rows that have a conservation status
    # 4 parks

df_species.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB


In [14]:
# Observations Data

In [15]:
# Numercial Data
    # Observations: Mean 142, Median 124, Min 9, Max 321
df_obs.describe()

observations
count  23296.000000
mean     142.287904
std       69.890532
min        9.000000
25%       86.000000
50%      124.000000
75%      195.000000
max      321.000000

In [16]:
# Categorical Data 
    # DUPLICATES: Scientific Name and Park Name 
     # Some Scientific Names are repeated as many as 12 times, all repeated at least 4
df_obs.scientific_name.value_counts()

scientific_name
Myotis lucifugus                        12
Puma concolor                           12
Hypochaeris radicata                    12
Holcus lanatus                          12
Streptopelia decaocto                   12
                                        ..
Packera dimorphophylla var. paysonii     4
Smilax bona-nox                          4
Chondestes grammacus                     4
Leymus triticoides                       4
Dichanthelium depauperatum               4
Name: count, Length: 5541, dtype: int64

In [17]:
df_obs.scientific_name.nunique()

5541

### Flagging Duplicates

1096 Rows where there are duplicated "scientific_name" & "park_name"

In [81]:
df_grouped = (df.groupby(["scientific_name", "park_name"]).common_names.count() > 1)

In [78]:
(df.groupby(["scientific_name", "park_name"]).common_names.count() > 1).reset_index

<bound method Series.reset_index of scientific_name                      park_name                          
Abies bifolia                        Bryce National Park                    False
                                     Great Smoky Mountains National Park    False
                                     Yellowstone National Park              False
                                     Yosemite National Park                 False
Abies concolor                       Bryce National Park                    False
                                                                            ...  
Zygodon viridissimus                 Yosemite National Park                 False
Zygodon viridissimus var. rupestris  Bryce National Park                    False
                                     Great Smoky Mountains National Park    False
                                     Yellowstone National Park              False
                                     Yosemite National Park            

In [82]:
df_group = pd.DataFrame(df_grouped)

In [85]:
df_group[df_group.common_names == True].index

MultiIndex([('Agrostis capillaris',                 'Bryce National Park'),
            ('Agrostis capillaris', 'Great Smoky Mountains National Park'),
            ('Agrostis capillaris',           'Yellowstone National Park'),
            ('Agrostis capillaris',              'Yosemite National Park'),
            (  'Agrostis gigantea',                 'Bryce National Park'),
            (  'Agrostis gigantea', 'Great Smoky Mountains National Park'),
            (  'Agrostis gigantea',           'Yellowstone National Park'),
            (  'Agrostis gigantea',              'Yosemite National Park'),
            ( 'Agrostis mertensii',                 'Bryce National Park'),
            ( 'Agrostis mertensii', 'Great Smoky Mountains National Park'),
            ...
            (      'Vulpia myuros',           'Yellowstone National Park'),
            (      'Vulpia myuros',              'Yosemite National Park'),
            (   'Vulpia octoflora',                 'Bryce National Park

In [86]:
df[df.scientific_name == "Zizia aptera"]

scientific_name                            park_name  observations  \
20208    Zizia aptera            Yellowstone National Park           257   
20209    Zizia aptera            Yellowstone National Park           257   
20210    Zizia aptera  Great Smoky Mountains National Park           105   
20211    Zizia aptera  Great Smoky Mountains National Park           105   
20212    Zizia aptera               Yosemite National Park           123   
20213    Zizia aptera               Yosemite National Park           123   
20214    Zizia aptera                  Bryce National Park           105   
20215    Zizia aptera                  Bryce National Park           105   
20216    Zizia aptera  Great Smoky Mountains National Park            41   
20217    Zizia aptera  Great Smoky Mountains National Park            41   
20218    Zizia aptera                  Bryce National Park           112   
20219    Zizia aptera                  Bryce National Park           112   
20220    Zizia aptera               Yosemite National Park           129   
20221    Zizia aptera               Yosemite National Park           129   
20222    Zizia aptera            Yellowstone National Park           269   
20223    Zizia aptera            Yellowstone National Park           269   

             category                                       common_names  \
20208  Vascular Plant                                  Golden Alexanders   
20209  Vascular Plant  Heartleaf Alexanders, Heart-Leaf Alexanders, M...   
20210  Vascular Plant                                  Golden Alexanders   
20211  Vascular Plant  Heartleaf Alexanders, Heart-Leaf Alexanders, M...   
20212  Vascular Plant                                  Golden Alexanders   
20213  Vascular Plant  Heartleaf Alexanders, Heart-Leaf Alexanders, M...   
20214  Vascular Plant                                  Golden Alexanders   
20215  Vascular Plant  Heartleaf Alexanders, Heart-Leaf Alexanders, M...   
20216  Vascular Plant                                  Golden Alexanders   
20217  Vascular Plant  Heartleaf Alexanders, Heart-Leaf Alexanders, M...   
20218  Vascular Plant                                  Golden Alexanders   
20219  Vascular Plant  Heartleaf Alexanders, Heart-Leaf Alexanders, M...   
20220  Vascular Plant                                  Golden Alexanders   
20221  Vascular Plant  Heartleaf Alexanders, Heart-Leaf Alexanders, M...   
20222  Vascular Plant                                  Golden Alexanders   
20223  Vascular Plant  Heartleaf Alexanders, Heart-Leaf Alexanders, M...   

      conservation_status  
20208                 NaN  
20209                 NaN  
20210                 NaN  
20211                 NaN  
20212                 NaN  
20213                 NaN  
20214                 NaN  
20215                 NaN  
20216                 NaN  
20217                 NaN  
20218                 NaN  
20219                 NaN  
20220                 NaN  
20221                 NaN  
20222                 NaN  
20223                 NaN

In [59]:
df[(df.groupby(["scientific_name", "park_name"]).common_names.count() > 1) == True]

/var/folders/75/8cznfb4d2j56mrdwqyh36s3w0000gn/T/ipykernel_11976/95369780.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df[(df.groupby(["scientific_name", "park_name"]).common_names.count() > 1) == True]


IndexingError: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match).

In [22]:
df_obs[(df_obs.scientific_name == "Myotis lucifugus") & (df_obs.park_name == "Great Smoky Mountains National Park")]

scientific_name                            park_name  observations
1603   Myotis lucifugus  Great Smoky Mountains National Park            57
8308   Myotis lucifugus  Great Smoky Mountains National Park            59
11395  Myotis lucifugus  Great Smoky Mountains National Park            58

Observations of same species in same park may have multiple values (57, 58, 59)

### 

In [88]:
# Categorical Park Names
    # Four Parks Represented Equally (Smoky Mountains, Bryce, Yellowstone, Yosemite)
df.park_name.value_counts(dropna=False)

park_name
Great Smoky Mountains National Park    6408
Yosemite National Park                 6408
Yellowstone National Park              6408
Bryce National Park                    6408
Name: count, dtype: int64

In [89]:
df.category.value_counts(dropna=False)

category
Vascular Plant       19560
Bird                  2364
Nonvascular Plant     1332
Mammal                1200
Fish                   524
Amphibian              328
Reptile                324
Name: count, dtype: int64

### Potential Opportunity to use Regex 

Some common_names include a duplicate name, with more names specified

In [90]:
df.common_names

0        Purple Vetch, Reddish Tufted Vetch
1        Purple Vetch, Reddish Tufted Vetch
2        Purple Vetch, Reddish Tufted Vetch
3        Purple Vetch, Reddish Tufted Vetch
4                             American Mink
                        ...                
25627                             Wolverine
25628        Green Panic-Grass, Panic Grass
25629        Green Panic-Grass, Panic Grass
25630        Green Panic-Grass, Panic Grass
25631        Green Panic-Grass, Panic Grass
Name: common_names, Length: 25632, dtype: object

In [91]:
df.scientific_name

0                Vicia benghalensis
1                Vicia benghalensis
2                Vicia benghalensis
3                Vicia benghalensis
4                    Neovison vison
                    ...            
25627                     Gulo gulo
25628    Dichanthelium depauperatum
25629    Dichanthelium depauperatum
25630    Dichanthelium depauperatum
25631    Dichanthelium depauperatum
Name: scientific_name, Length: 25632, dtype: object

## Group by Endangered 

Count the species from each category

## Species Info Data

In [26]:
df_species.head()

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   
3   Mammal                     Ovis aries   
4   Mammal                 Cervus elaphus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN  
3  Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)                 NaN  
4                                      Wapiti Or Elk                 NaN

In [27]:
df_species.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB


In [32]:
df_species.scientific_name.nunique()

5541

In [95]:
df.tail(50)

scientific_name  \
25582  Croton glandulosus var. septentrionalis   
25583  Croton glandulosus var. septentrionalis   
25584                    Desmodium pauciflorum   
25585                    Desmodium pauciflorum   
25586                    Desmodium pauciflorum   
25587                    Desmodium pauciflorum   
25588   Ranunculus alismifolius var. hartwegii   
25589   Ranunculus alismifolius var. hartwegii   
25590   Ranunculus alismifolius var. hartwegii   
25591   Ranunculus alismifolius var. hartwegii   
25592                   Symphyotrichum boreale   
25593                   Symphyotrichum boreale   
25594                   Symphyotrichum boreale   
25595                   Symphyotrichum boreale   
25596                       Aconitum uncinatum   
25597                       Aconitum uncinatum   
25598                       Aconitum uncinatum   
25599                       Aconitum uncinatum   
25600                           Stachys albens   
25601                           Stachys albens   
25602                           Stachys albens   
25603                           Stachys albens   
25604                               Sium suave   
25605                               Sium suave   
25606                               Sium suave   
25607                               Sium suave   
25608                     Claytonia nevadensis   
25609                     Claytonia nevadensis   
25610                     Claytonia nevadensis   
25611                     Claytonia nevadensis   
25612                         Listera borealis   
25613                         Listera borealis   
25614                         Listera borealis   
25615                         Listera borealis   
25616   Trifolium gracilentum var. gracilentum   
25617   Trifolium gracilentum var. gracilentum   
25618   Trifolium gracilentum var. gracilentum   
25619   Trifolium gracilentum var. gracilentum   
25620                    Ranunculus hebecarpus   
25621                    Ranunculus hebecarpus   
25622                    Ranunculus hebecarpus   
25623                    Ranunculus hebecarpus   
25624                                Gulo gulo   
25625                                Gulo gulo   
25626                                Gulo gulo   
25627                                Gulo gulo   
25628               Dichanthelium depauperatum   
25629               Dichanthelium depauperatum   
25630               Dichanthelium depauperatum   
25631               Dichanthelium depauperatum   

                                 park_name  observations        category  \
25582               Yosemite National Park           172  Vascular Plant   
25583            Yellowstone National Park           264  Vascular Plant   
25584  Great Smoky Mountains National Park            76  Vascular Plant   
25585                  Bryce National Park           101  Vascular Plant   
25586            Yellowstone National Park           219  Vascular Plant   
25587               Yosemite National Park           152  Vascular Plant   
25588  Great Smoky Mountains National Park           100  Vascular Plant   
25589            Yellowstone National Park           239  Vascular Plant   
25590                  Bryce National Park            95  Vascular Plant   
25591               Yosemite National Park           150  Vascular Plant   
25592  Great Smoky Mountains National Park            79  Vascular Plant   
25593                  Bryce National Park            66  Vascular Plant   
25594               Yosemite National Park           139  Vascular Plant   
25595            Yellowstone National Park           220  Vascular Plant   
25596               Yosemite National Park           142  Vascular Plant   
25597  Great Smoky Mountains National Park            77  Vascular Plant   
25598                  Bryce National Park           115  Vascular Plant   
25599            Yellowstone National Park           248  Vascular Plant   
25600               Yosemite Nationa

df.scientific_name.nunique()

# Cleaning Observation Data

In [98]:
df_obs

scientific_name  \
0                           Vicia benghalensis   
1                               Neovison vison   
2                            Prunus subcordata   
3                         Abutilon theophrasti   
4                     Githopsis specularioides   
...                                        ...   
23291                     Croton monanthogynus   
23292                 Otospermophilus beecheyi   
23293  Heterotheca sessiliflora ssp. echioides   
23294                     Dicranella rufescens   
23295                           Cucurbita pepo   

                                 park_name  observations  
0      Great Smoky Mountains National Park            68  
1      Great Smoky Mountains National Park            77  
2                   Yosemite National Park           138  
3                      Bryce National Park            84  
4      Great Smoky Mountains National Park            85  
...                                    ...           ...  
23291               Yosemite National Park           173  
23292                  Bryce National Park           130  
23293                  Bryce National Park           140  
23294               Yosemite National Park           171  
23295               Yosemite National Park           164  

[23296 rows x 3 columns]

In [104]:
round(df_obs.groupby(["scientific_name", "park_name"]).observations.median())

scientific_name                      park_name                          
Abies bifolia                        Bryce National Park                    109.0
                                     Great Smoky Mountains National Park     72.0
                                     Yellowstone National Park              215.0
                                     Yosemite National Park                 136.0
Abies concolor                       Bryce National Park                     83.0
                                                                            ...  
Zygodon viridissimus                 Yosemite National Park                 159.0
Zygodon viridissimus var. rupestris  Bryce National Park                    102.0
                                     Great Smoky Mountains National Park    102.0
                                     Yellowstone National Park              237.0
                                     Yosemite National Park                 210.0
Name: observations, Lengt

In [ ]:
df_obs.observations.transform(lambda x: round(df_obs.groupby(["scientific_name", "park_name"]).observations.median()))

In [ ]:
df

In [113]:
test_array = df_obs.groupby(["scientific_name", "park_name"]).observations.transform(lambda x: round(x.median()))

In [107]:
df.observations = df.groupby(["scientific_name", "park_name"]).observations.transform(lambda x: round(x.median()))

0         68
1        148
2        247
3        104
4         77
        ... 
25627    157
25628     88
25629     80
25630    255
25631    162
Name: observations, Length: 25632, dtype: int64

In [115]:
df_obs.observations

0         68
1         77
2        138
3         84
4         85
        ... 
23291    173
23292    130
23293    140
23294    171
23295    164
Name: observations, Length: 23296, dtype: int64

In [114]:
test_array

0         68
1         77
2        138
3         84
4         85
        ... 
23291    173
23292    130
23293    140
23294    171
23295    164
Name: observations, Length: 23296, dtype: int64

In [112]:
(df_obs.groupby(["scientific_name", "park_name"]).observations.count() > 1).sum()

1096

In [119]:
df_obs.duplicated().sum()

15

In [120]:
df_obs.observations = df_obs.groupby(["scientific_name", "park_name"]).observations.transform(lambda x: round(x.median()))

In [122]:
df_obs.shape

(23296, 3)

In [123]:
df_obs.drop_duplicates(inplace=True)

In [133]:
# Now no combination of Scientific Name and Park Name are observed multuple times 
# Each scientific name and park name occur one time
(df_obs.groupby(["scientific_name", "park_name"]).observations.count() > 1).sum()

0

In [125]:
df_obs.duplicated().sum()

0

Now each species has 1 observation per park 
(if previously multiple, filled in with the median of the multiple entries)

In [138]:
df_obs.groupby("scientific_name").observations.count()

scientific_name
Abies bifolia                          4
Abies concolor                         4
Abies fraseri                          4
Abietinella abietina                   4
Abronia ammophila                      4
                                      ..
Zonotrichia leucophrys gambelii        4
Zonotrichia leucophrys oriantha        4
Zonotrichia querula                    4
Zygodon viridissimus                   4
Zygodon viridissimus var. rupestris    4
Name: observations, Length: 5541, dtype: int64

In [139]:
df_obs.groupby("park_name").observations.count()

park_name
Bryce National Park                    5541
Great Smoky Mountains National Park    5541
Yellowstone National Park              5541
Yosemite National Park                 5541
Name: observations, dtype: int64

In [143]:
df_obs.groupby("scientific_name").observations.count() 

scientific_name
Abies bifolia                          4
Abies concolor                         4
Abies fraseri                          4
Abietinella abietina                   4
Abronia ammophila                      4
                                      ..
Zonotrichia leucophrys gambelii        4
Zonotrichia leucophrys oriantha        4
Zonotrichia querula                    4
Zygodon viridissimus                   4
Zygodon viridissimus var. rupestris    4
Name: observations, Length: 5541, dtype: int64

5541 different species 

Same amount in Observations and Species Dataframes

In [132]:
df_obs[df_obs.scientific_name.str.contains('Zygodon')]

scientific_name  \
3649                  Zygodon viridissimus   
5015                  Zygodon viridissimus   
5539                  Zygodon viridissimus   
6879   Zygodon viridissimus var. rupestris   
12775  Zygodon viridissimus var. rupestris   
16130                 Zygodon viridissimus   
16237  Zygodon viridissimus var. rupestris   
20040  Zygodon viridissimus var. rupestris   

                                 park_name  observations  
3649   Great Smoky Mountains National Park            71  
5015             Yellowstone National Park           270  
5539                Yosemite National Park           159  
6879             Yellowstone National Park           237  
12775                  Bryce National Park           102  
16130                  Bryce National Park           100  
16237               Yosemite National Park           210  
20040  Great Smoky Mountains National Park           102

# Cleaning Species Data

NOTE Multiple Common names 

In [140]:
df_species.head()

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   
3   Mammal                     Ovis aries   
4   Mammal                 Cervus elaphus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN  
3  Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)                 NaN  
4                                      Wapiti Or Elk                 NaN

In [144]:
(df_species.groupby(["scientific_name"]).category.count() > 1).sum()

274

In [148]:
(df_species.groupby(["scientific_name"]).common_names.count() > 1).sum()

274

274 Duplicated Scientific Names in Species

Common Names are source of duplicates

In [160]:
repeated_common_names = np.where(df_species.groupby(["scientific_name"]).common_names.count() > 1)

In [164]:
species_w_repeated_common_names = df_species.iloc[repeated_common_names].scientific_name
species_w_repeated_common_names

In [186]:
df_species[df_species.scientific_name.isin(species_w_repeated_common_names)].groupby(["scientific_name"]).common_names.count()

scientific_name
Aeronautes saxatalis                    1
Agastache urticifolia                   1
Agrostis oregonensis                    1
Allophyllum gilioides ssp. violaceum    1
Alnus rhombifolia                       1
                                       ..
Viola bakeri                            1
Viola porterana                         1
Vireo solitarius                        2
Whitneya dealbata                       1
Zigadenus venenosus var. gramineus      1
Name: common_names, Length: 274, dtype: int64

In [158]:
df_species[df_species.scientific_name == "Pandion haliaetus"]

category    scientific_name            common_names conservation_status
104      Bird  Pandion haliaetus                  Osprey  Species of Concern
3081     Bird  Pandion haliaetus  Osprey, Western Osprey  Species of Concern

In [ ]:
df_species.common_names = df_species.groupby(["scientific_name"]).common_names.transform(lambda x: len(x).max()))

In [181]:
testing = df_species.groupby(["scientific_name"]).common_names.transform(lambda x: x)

In [183]:
testing.duplicated().sum()

320

In [184]:
testing.shape

(5824,)

In [185]:
5824 - 320

5504

In [188]:
df_species[df_species.scientific_name == "Viola bakeri"]

category scientific_name    common_names conservation_status
5378  Vascular Plant    Viola bakeri  Baker's Violet                 NaN

In [193]:
pd.set_option("display.max_rows", 300)

In [206]:
#(pd.DataFrame(df_species[df_species.scientific_name.duplicated()].groupby("scientific_name")))

In [207]:
df_species[df_species.scientific_name == "Canis lupus"]

category scientific_name     common_names conservation_status
8      Mammal     Canis lupus        Gray Wolf          Endangered
3020   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
4448   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered

## IMPORTANT CODE

Drop duplicated scienfitic names, where conservation status is not null

The only change is that common_names may not include as many, which is acceptable for this purpose. Does not give information about endangered specises

In [214]:
duplicated_not_endangered = df_species[df_species.scientific_name.duplicated() & (df_species.conservation_status.isnull())].index
df_species.drop(duplicated_not_endangered, inplace = True)

Check if all duplicated species have the same conservation status for all their rows

Note the species that have different conservation statuses
* Canis lupus (keep "Gray Wolf, Wolf", 	"Endangered")
* Oncorhynchus mykiss (threatened vs nan, keep threatened)


Remove Specific Rows

In [232]:
df_species.head(10)

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   
3   Mammal                     Ovis aries   
4   Mammal                 Cervus elaphus   
5   Mammal         Odocoileus virginianus   
6   Mammal                     Sus scrofa   
7   Mammal                  Canis latrans   
8   Mammal                    Canis lupus   
9   Mammal                    Canis rufus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN  
3  Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)                 NaN  
4                                      Wapiti Or Elk                 NaN  
5                                  White-Tailed Deer                 NaN  
6                                Feral Hog, Wild Pig                 NaN  
7                                             Coyote  Species of Concern  
8                                          Gray Wolf          Endangered  
9                                           Red Wolf          Endangered

# IMPORTANT CODE

In [236]:
# First 8 and 3020 correspond to Wolf and 560 corresponds to Oncorhynchus mykiss
df_species.drop([8,3020,560],inplace=True)

In [240]:
duplicated_endangered = df_species[df_species.scientific_name.duplicated()].index
df_species.drop(duplicated_endangered, inplace = True)

In [229]:
df_species[df_species.scientific_name == "Canis lupus"]

category scientific_name     common_names conservation_status
8      Mammal     Canis lupus        Gray Wolf          Endangered
3020   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
4448   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered

In [234]:
df_species[df_species.scientific_name == "Oncorhynchus mykiss"]

category      scientific_name   common_names conservation_status
560      Fish  Oncorhynchus mykiss  Rainbow Trout                 NaN
3283     Fish  Oncorhynchus mykiss  Rainbow Trout          Threatened

In [244]:
df_species.shape[0] * 4

22164

In [245]:
df_obs.shape

(22164, 3)

In [241]:
df_species[df_species.scientific_name.duplicated()]

Empty DataFrame
Columns: [category, scientific_name, common_names, conservation_status]
Index: []

In [239]:
df_species[df_species.scientific_name.duplicated()]

category            scientific_name  \
3035   Mammal           Eptesicus fuscus   
3037   Mammal  Lasionycteris noctivagans   
3042   Mammal           Myotis lucifugus   
3081     Bird          Pandion haliaetus   
3150     Bird                Gavia immer   
3185     Bird            Riparia riparia   
4457   Mammal              Taxidea taxus   
4465   Mammal        Myotis californicus   
4467   Mammal           Myotis lucifugus   
4564     Bird      Nycticorax nycticorax   

                                           common_names conservation_status  
3035                       Big Brown Bat, Big Brown Bat  Species of Concern  
3037               Silver-Haired Bat, Silver-Haired Bat  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
3081                             Osprey, Western Osprey  Species of Concern  
3150  Common Loon, Great Northern Diver, Great North...  Species of Concern  
3185                          Bank Swallow, Sand Martin  Species of Concern  
4457                                             Badger  Species of Concern  
4465                                  California Myotis  Species of Concern  
4467                                Little Brown Myotis  Species of Concern  
4564                          Black-Crowned Night Heron  Species of Concern

In [215]:
df_species[df_species.scientific_name.duplicated()]

category            scientific_name  \
3020   Mammal                Canis lupus   
3035   Mammal           Eptesicus fuscus   
3037   Mammal  Lasionycteris noctivagans   
3042   Mammal           Myotis lucifugus   
3081     Bird          Pandion haliaetus   
3150     Bird                Gavia immer   
3185     Bird            Riparia riparia   
3283     Fish        Oncorhynchus mykiss   
4448   Mammal                Canis lupus   
4457   Mammal              Taxidea taxus   
4465   Mammal        Myotis californicus   
4467   Mammal           Myotis lucifugus   
4564     Bird      Nycticorax nycticorax   

                                           common_names conservation_status  
3020                                    Gray Wolf, Wolf         In Recovery  
3035                       Big Brown Bat, Big Brown Bat  Species of Concern  
3037               Silver-Haired Bat, Silver-Haired Bat  Species of Concern  
3042  Little Brown Bat, Little Brown Myotis, Little ...  Species of Concern  
3081                             Osprey, Western Osprey  Species of Concern  
3150  Common Loon, Great Northern Diver, Great North...  Species of Concern  
3185                          Bank Swallow, Sand Martin  Species of Concern  
3283                                      Rainbow Trout          Threatened  
4448                                    Gray Wolf, Wolf          Endangered  
4457                                             Badger  Species of Concern  
4465                                  California Myotis  Species of Concern  
4467                                Little Brown Myotis  Species of Concern  
4564                          Black-Crowned Night Heron  Species of Concern

In [205]:
df_species[df_species.scientific_name.duplicated()]

category                     scientific_name  \
3017          Mammal                      Cervus elaphus   
3019          Mammal              Odocoileus virginianus   
3020          Mammal                         Canis lupus   
3022          Mammal                       Puma concolor   
3025          Mammal                    Lutra canadensis   
3028          Mammal                       Mustela vison   
3030          Mammal                       Procyon lotor   
3031          Mammal                    Ursus americanus   
3035          Mammal                    Eptesicus fuscus   
3037          Mammal           Lasionycteris noctivagans   
3038          Mammal                   Lasiurus cinereus   
3042          Mammal                    Myotis lucifugus   
3050          Mammal                   Castor canadensis   
3056          Mammal                  Ondatra zibethicus   
3057          Mammal              Peromyscus maniculatus   
3071          Mammal                      Sorex cinereus   
3072          Mammal                          Sorex hoyi   
3075          Mammal                     Sorex palustris   
3081            Bird                   Pandion haliaetus   
3088            Bird                     Branta bernicla   
3091            Bird                   Chen caerulescens   
3138            Bird                       Columba livia   
3140            Bird               Streptopelia decaocto   
3144            Bird                    Alectoris chukar   
3150            Bird                         Gavia immer   
3169            Bird                   Passerella iliaca   
3184            Bird            Petrochelidon pyrrhonota   
3185            Bird                     Riparia riparia   
3191            Bird              Dumetella carolinensis   
3192            Bird                   Mimus polyglottos   
3194            Bird                    Anthus rubescens   
3214            Bird                 Polioptila caerulea   
3217            Bird                    Sturnus vulgaris   
3231            Bird                    Vireo solitarius   
3233            Bird                 Butorides virescens   
3252            Bird                      Megascops asio   
3267       Amphibian                    Rana catesbeiana   
3283            Fish                 Oncorhynchus mykiss   
3285            Fish               Salvelinus fontinalis   
3295  Vascular Plant                    Lemna valdiviana   
3298  Vascular Plant                    Elodea nuttallii   
3302  Vascular Plant             Potamogeton amplifolius   
3303  Vascular Plant               Potamogeton epihydrus   
3309  Vascular Plant                 Potamogeton nodosus   
3328  Vascular Plant                         Carum carvi   
3330  Vascular Plant                    Conium maculatum   
3332  Vascular Plant                       Daucus carota   
3343  Vascular Plant                    Pastinaca sativa   
3345  Vascular Plant                Sanicula marilandica   
3347  Vascular Plant                        Zizia aptera   
3354  Vascular Plant               Asparagus officinalis   
3367  Vascular Plant            Corallorrhiza wisteriana   
3386  Vascular Plant             Ambrosia artemisiifolia   
3405  Vascular Plant                       Arctium minus   
3423  Vascular Plant                       Bidens cernua   
3425  Vascular Plant                 Carduus acanthoides   
3432  Vascular Plant                   Cichorium intybus   
3433  Vascular Plant                     Cirsium arvense   
3439  Vascular Plant                   Conyza canadensis   
3466  Vascular Plant                  Erigeron strigosus   
3480  Vascular Plant                   Helianthus annuus   
3484  Vascular Plant               Hieracium caespitosum   
3488  Vascular Plant                Hypochaeris radicata   
3491  Vascular Plant                     Lactuca biennis   
3492  Vascular Plant                    Lactuca serriola   
3526  Vascular Plant                   Solidago giga

In [192]:
5824 - 283

5541

Get rid of duplicates in scientific name

Choose the option with longest common_name or worst status of endangered

# Merge Cleaned Data

In [246]:
# Merge Data Together on Shared Column "scientific_name"
df = df_obs.merge(df_species, on="scientific_name")

In [247]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22164 entries, 0 to 22163
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   scientific_name      22164 non-null  object
 1   park_name            22164 non-null  object
 2   observations         22164 non-null  int64 
 3   category             22164 non-null  object
 4   common_names         22164 non-null  object
 5   conservation_status  716 non-null    object
dtypes: int64(1), object(5)
memory usage: 1.0+ MB


In [251]:
(df.groupby(by=["scientific_name", "park_name"]).category.count() > 1).sum()

0

# Clean Data!